stl/inc/numeric

```diff
@@ -70,31 +70,40 @@ inline constexpr bool _Plus_on_arithmetic_ranges_reduction_v = false;
 #endif // _STD_VECTORIZE_WITH_FLOAT_CONTROL
 
 template <class _InIt, class _Ty, class _BinOp>
-_NODISCARD _Ty reduce(const _InIt _First, const _InIt _Last, _Ty _Val, _BinOp _Reduce_op) {
+#ifdef __cpp_lib_is_constant_evaluated
+_NODISCARD _CONSTEXPR20 _Ty reduce(
+#else // ^^^ __cpp_lib_is_constant_evaluated / !__cpp_lib_is_constant_evaluated vvv
+_NODISCARD _Ty reduce(
+#endif // __cpp_lib_is_constant_evaluated
+    const _InIt _First, const _InIt _Last, _Ty _Val, _BinOp _Reduce_op) {
     // return commutative and associative reduction of _Val and [_First, _Last), using _Reduce_op
     _Adl_verify_range(_First, _Last);
     auto _UFirst      = _Get_unwrapped(_First);
     const auto _ULast = _Get_unwrapped(_Last);
     if constexpr (_Plus_on_arithmetic_ranges_reduction_v<_Unwrapped_t<const _InIt&>, _Ty, _BinOp>) {
-        (void) _Reduce_op; // TRANSITION, VSO-486357
-        return _Reduce_plus_arithmetic_ranges(_UFirst, _ULast, _Val);
-    } else {
-        for (; _UFirst != _ULast; ++_UFirst) {
-            _Val = _Reduce_op(_STD move(_Val), *_UFirst); // Requirement missing from N4713
+#ifdef __cpp_lib_is_constant_evaluated
+        if (!_STD is_constant_evaluated())
+#endif // __cpp_lib_is_constant_evaluated
+        {
+            (void) _Reduce_op; // TRANSITION, VSO-486357
+            return _Reduce_plus_arithmetic_ranges(_UFirst, _ULast, _Val);
         }
-
-        return _Val;
     }
+    for (; _UFirst != _ULast; ++_UFirst) {
+        _Val = _Reduce_op(_STD move(_Val), *_UFirst); // Requirement missing from N4713
+    }
+
+    return _Val;
 }
 
 template <class _InIt, class _Ty>
-_NODISCARD _Ty reduce(const _InIt _First, const _InIt _Last, _Ty _Val) {
+_NODISCARD _CONSTEXPR20 _Ty reduce(const _InIt _First, const _InIt _Last, _Ty _Val) {

```

Comment:
This _CONSTEXPR20 looks ill formed, no diagnostic required for implementations that don't have is_constant_evaluated.

Reply:
I believe, that should be alleviated with your `_CONSTEXPR_ICE` suggestion?

Reply:
Fix by define `_CONSTEXPR20_ICE`